lib/Doctrine/Common/DataFixtures/Sorter/TopologicalSorter.php

```diff
@@ -0,0 +1,198 @@
+     *
+     * @param string $hash
+     *
+     * @return bool
+     */
+    public function hasNode($hash)
+    {
+        return isset($this->nodeList[$hash]);
+    }
+
+    /**
+     * Adds a new dependency (edge) to the graph using their hashes.
+     *
+     * @param string $fromNode
+     * @param string $toNode
+     *
+     * @return void
+     */
+    public function addDependency($fromHash, $toHash)
+    {
+        $definition = $this->nodeList[$fromHash];
+
+        $definition->dependencyList[] = $toHash;
+    }
+
+    /**
+     * Sets the dependency list (edges) to the graph using their hashes.
+     *
+     * @param string $fromHash
+     * @param array  $dependencyList
+     *
+     * @return void
+     */
+    public function setDependencyList($fromHash, array $dependencyList)
+    {
+        $definition = $this->nodeList[$fromHash];
+
+        $definition->dependencyList = $dependencyList;
+    }
+
+    /**
+     * Return a valid order list of all current nodes.
+     * The desired topological sorting is the postorder of these searches.
+     *
+     * {@internal Highly performance-sensitive method.}
+     *
+     * @return array
+     */
+    public function sort()
+    {
+        foreach ($this->nodeList as $definition) {
+            if ($definition->state !== self::NOT_VISITED) {
+                continue;
+            }
+
+            $this->visit($definition);
+        }
+
+        $sortedList = $this->sortedNodeList;
+
+        $this->nodeList       = array();
+        $this->sortedNodeList = array();
+
+        return $sortedList;
+    }
+
+    /**
+     * Visit a given node definition for reordering.
+     *
+     * {@internal Highly performance-sensitive method.}
+     *
+     * @throws \RuntimeException
+     *
+     * @param \stdClass $definition
+     */
+    private function visit($definition)
+    {
+        $definition->state = self::IN_PROGRESS;
+
+        foreach ($definition->dependencyList as $dependency) {
+            if ( ! isset($this->nodeList[$dependency])) {
+                throw new \RuntimeException(
+                    sprintf(
+                        'Fixture "%s" has a dependency of fixture "%s", but it not listed to be loaded.',
+                        get_class($definition->value),
+                        $dependency
+                    )
+                );
+            }
+
+            $childDefinition = $this->nodeList[$dependency];
+
+            switch ($childDefinition->state) {
+                case self::VISITED:
+                    continue;
+
+                case self::IN_PROGRESS:
+                    $message = 'Graph contains cyclic dependency. An example of this problem would be the following: '
+                        . 'Class C has class B as its dependency. Then, class B has class A has its dependency. '
+                        . 'Finally, class A has class C as its dependency.';
+
+                    throw new \RuntimeException($message);
```

Comment:
Can we please throw a custom exception here (from the data-fixtures namespace, at least)?